### DISCUSSION TITLE
Run vscodium in browser

### DISCUSSION THREAD

**Discussion by 1kastner** [VSCodium/vscodium#1469] [General]
In some cases we want to code in a browser, e.g. when the code should be executed on a [JupyterHub](https://jupyterhub.readthedocs.io/en/stable/). However, currently the only option google provided me with was using [code-server](https://github.com/cdr/code-server) who are gathering the telemetries for themselves (see https://github.com/betatim/vscode-binder/issues/22 and mentioned inside for further information). I would prefer to do the same without sharing telemetry data.

So I wondered whether it would be easy to not run vscodium on a remote server and being served the UI to a modern browser of my choice. To be even more specific, I would love [this](https://github.com/betatim/vscode-binder) to work with vscodium but I am a bit lost in this repository on how to achieve this.

Thank you so much for this project and some inspiration!

**Comment by GitMensch** [comment]
Only related by the topic, but did you tried to use Theia for the browser parts? Can you explain where the difference between this and your suggested option is?

**Comment by 1kastner** [comment]
I haven't tried out Theia and it seems like [it should already work](https://jupyter-server-proxy.readthedocs.io/en/latest/convenience/packages/theia.html)! As far as I have read [in this blog article](https://eclipsesource.com/de/blogs/2019/12/06/the-eclipse-theia-ide-vs-vs-code/), Theia does not yet fully support the API for extensions that do work with vscode/vscodium. Furthermore, [on their github page](https://github.com/eclipse-theia/theia) they suggest to run the editor in a separate docker container instead of offering OS-native installers like [vscodium](https://github.com/vscodium/vscodium/releases). That does change a bit how a binder example like [here](https://github.com/betatim/vscode-binder) can be created with smallest effort. I would prefer just to tweak a few lines and plug together my system that will work over several versions. That e.g. forbids to just copy a docker file that specifies one IDE (see e.g. the [theia Python IDE](https://github.com/theia-ide/theia-apps/blob/master/theia-python-docker/Dockerfile)) by hand and translate this into a Shell script for installation. Tools like [dockerfile-to-shell-script](https://github.com/thatkevin/dockerfile-to-shell-script/blob/master/docker_to_sh.sh) or [dkr2sh](https://github.com/devgateway/dkr2sh) both describe themselves as naive and one needs to be careful with the result.

**Comment by cothan** [comment]
If you have student github account, you can definitely try gitpod.io https://gitpod.io/ 

Disclaimer: I don't have any relation with gitpod, I'm coding on RPi4 and find out that using gitpod make it easier, faster UI for me to code.

**Comment by 1kastner** [comment]
Thank you very much for pointing that one out. I haven't used it for quite a while. But it seems that it is open source now so that I could even host the IDE myself - see [its git repo](https://github.com/gitpod-io/gitpod). Under the hood [they use Theia](https://www.gitpod.io/docs/ide/). So mainly gitpod adds all the options for integrations into different platform such as GitHub, GitLab etc.

**Comment by Felx-B** [comment]
Hi there,
I compiled VSCode for browser in this repo : https://github.com/Felx-B/vscode-web
It is a full browser version of VSCode like [official playground](https://vscode-web-test-playground.azurewebsites.net/)

This way you should be able to write your own extension with FileSystemProvider and remote execution.

**Comment by 1kastner** [comment]
That is great news! Did you check whether telemetry data is sent? I would love to give it a shot whether it can be integrated into MyBinder like it has been done at https://github.com/betatim/vscode-binder

**Reply by GitMensch** [reply]
Just checking @1kastner - as "running vscodium in the browser" is obviously possible now did you gave that integration a try already?

**Comment by Felx-B** [comment]
By default there is no telemetry.
You can choose to add your own by settings some properties in `product.json` file.
In this file you can setup any key of [IProductConfiguration](https://github.com/microsoft/vscode/blob/300effb0c8f5722cee4ae08578705438bdaff075/src/vs/platform/product/common/productService.ts#L34)

It seems you can only use ApplicationInsight telemetry by setting these keys : 
```ts
readonly enableTelemetry?: boolean;
readonly aiConfig?: {
		readonly asimovKey: string;
	};
```

**Comment by ghost** [comment]
so, is there a definitive answer for a web version of vscodium?

**Reply by bitsnaps** [reply]
AFAIK, the [code-server](https://github.com/cdr/code-server) as @1kastner mentioned seems to be the best option for now, you may need to tweak it a little bit to disable the [telemetry](https://code.visualstudio.com/docs/getstarted/telemetry).

**Comment by Kreijstal** [comment]
Is there no codium-server?

**Reply by Kreijstal** [reply]
Think about it you are not running electron, you just use your browser tab and codium should be a server, the best of both worlds, yeah you're exposing a port? I mean maybe you want to expose it to work remotely

**Reply by Cyberbeni** [reply]
`codium serve-web`

There is no official/popular docker container for it yet.

**Reply by nick-walt** [reply]
This one? 
`vscodium-reh-web-linux-x64-1.106.37943`
https://github.com/VSCodium/vscodium/blob/master/docs/others.md#reh

https://github.com/VSCodium/vscodium/releases
Web Host

Surprisingly there are no instructions for installation and setup.

**Reply by daiyam** [reply]
Yes, `reh-web` has only the server to run VSCodium in the browser.

The regular IDE also contains it (launched with `codium serve-web`)

**Reply by nick-walt** [reply]
Ok, thanks. Is Web Host server also compiled directly from a VSCode source or is it a custom VSCodium project? There does not appear to be a VSCode equivalent to VSCodium Web Host server.

VSCode for the Web exists but appears to be entirely cloud based similar to Codespaces but extremely limited. Visual Studio Code for the Web connects to https://vscode.dev

Does VSCodium Web Host function more like Remote SSH, where the browser-run GUI Editor functions similar to the full Desktop, or like https://vscode.dev/ ?

**Reply by daiyam** [reply]
> Ok, thanks. Is Web Host server also compiled directly from a VSCode source or is it a custom VSCodium project? There does not appear to be a VSCode equivalent to VSCodium Web Host server.

It's from a VSCode source. (https://github.com/search?q=repo%3Amicrosoft%2Fvscode%20reh-web&type=code)

> Does VSCodium Web Host function more like Remote SSH, where the browser-run GUI Editor functions similar to the full Desktop, or like https://vscode.dev/ ?

More like vscode.dev.

**Reply by nick-walt** [reply]
Ok, thank you.

**Comment by GitMensch** [comment]
Shouldn't this discussion be solved as resolved, now that vscodium-reh-web is available (since some time)?
